xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -49,169 +48,223 @@
+      // We assume if edsServiceName is null, it will always be null in later resolver updates;
+      // and if edsServiceName is not null, it will always be not null.
+      if (edsServiceName == null) {
+        edsServiceName = lookasideLbHelper.getAuthority();
+      }
 
-            @Override
-            public void handleNameResolutionError(Status error) {}
+      LoadBalancerProvider clusterEndpointsLoadBalancer =
+          new ClusterEndpointsBalancerProvider(edsServiceName);
+      switchingLoadBalancer.switchTo(clusterEndpointsLoadBalancer);
+    }
+    resolvedAddresses = resolvedAddresses.toBuilder()
+        .setAttributes(attributes.toBuilder().discard(ATTR_LOAD_BALANCING_CONFIG).build())
+        .setLoadBalancingPolicyConfig(newXdsConfig)
+        .build();
+    switchingLoadBalancer.handleResolvedAddresses(resolvedAddresses);
+    this.xdsConfig = newXdsConfig;
+
+    // TODO(zdapeng): If lrsServerName in XdsConfig is changed, call xdsClient.reportClientStats()
+    //     and/or xdsClient.cancelClientStatsReport().
+  }
 
-            @Override
-            public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
-              if (xdsClient == null) {
-                ManagedChannel channel = initLbChannel(helper, newBalancerName, channelCredsList);
-                xdsClient  = new XdsComms2(
-                    channel, helper, new ExponentialBackoffPolicy.Provider(),
-                    GrpcUtil.STOPWATCH_SUPPLIER, node);
-                LoadStatsStore loadStatsStore = new LoadStatsStoreImpl();
-                localityStore = localityStoreFactory.newLocalityStore(
-                    helper, lbRegistry, loadStatsStore);
-                // TODO(zdapeng): Use XdsClient to do Lrs directly.
-                lrsClient = loadReportClientFactory.createLoadReportClient(
-                    channel, helper, new ExponentialBackoffPolicy.Provider(),
-                    loadStatsStore);
-                final LoadReportCallback lrsCallback =
-                    new LoadReportCallback() {
-                      @Override
-                      public void onReportResponse(long reportIntervalNano) {
-                        localityStore.updateOobMetricsReportInterval(reportIntervalNano);
-                      }
-                    };
-
-                EndpointWatcher endpointWatcher =
-                    new EndpointWatcherImpl(lrsClient, lrsCallback, localityStore);
-                xdsClient.watchEndpointData(node.getCluster(), endpointWatcher);
-              }
-            }
+  @Override
+  public void handleNameResolutionError(Status error) {
+    channelLogger.log(ChannelLogLevel.ERROR, "Name resolution error: '%s'", error);
+    // Go into TRANSIENT_FAILURE if we have not yet received any endpoint update. Otherwise,
+    // we keep running with the data we had previously.
+    if (endpointWatcher == null) {
```

Comment:
Trying to match the code with the comment: why does `endpointWatcher == null` imply "we have not yet received any endpoint update"?

Reply:
If `endpointWatcher == null` we don't even have the ability to receive endpoint update, because that's before `handleResolvedAddresses()` successfully. If `endpointWatcher != null`, we further check `if (!endpointWatcher.firstEndpointUpdateReceived)` which implies "we have not yet received any endpoint update"